# CEE/BIO 6720: Laboratory 3
## Comparative metagenomics with assemblies (part 2)


## Purpose: 

In this exercise, we'll be analyzing a timeseries dataset captured during a sewage spill in a drinking water reservoir. We will use assembly-dependent methods to track sewage-associated populations over time. Students can expect to become familiar with topics in:

1. Assembly and binning
2. Annotation and taxonomic classification
3. Determining the relative abundance of populations across timeseries

The sewage was sequenced in three biological replicates, labelled as `sewageA`,`sewageB`, and `sewageC`. We will assemble and bin these metagenomes then track the resulting MAGs across the spill at 4 timepoints to examine the underlying microbial dynamics. The time points provided to you are four separate metagenomes collected from the impacted drinking water reservoir, labelled as so:

`hour000`,`hour024`,`hour096` and `hour168`

## Outline: 

•	Step 0: Set up your workspace and retrieve software.

•	Step 1: Trim raw reads with `fastp`

•	Step 2: Assemble reads with `metaSPAdes`

•	Step 3: Bin data with `maxbin` and `metabat`

•	Step 4: Check and annotate bins 

•	Step 5: Taxonomically classify good quality bins

•	Step 6: Ascertain their relative abundance 


# Instructions:

## **Step 0: Workspace preparation** 

1.  Copy `shared/lab03` to your `scratch` directory (`cp -r ~/shared/lab03 ~/scratch/`)
2.  Clean up `micromamba` archives and packages:
```
micromamba clean --all -y
```
3.  Create an environment with the software we need:
```
# if your command line can't recognize micromamba, try: source ~/.bashrc
micromamba create -p ~/scratch/lab03/env
micromamba activate ~/scratch/lab03/env
micromamba install fastp spades bakta coverm maxbin2 gtdbtk checkm2 metabat2
```

## **Step 1: Read trimming** 
1.  Use `step01.sbatch` to trim all the metagenomes provided to you (both the sewage and timeseries metagenomes).

## **Step 2: Metagenomic assembly** 
1.  Use `step02.sbatch` to assemble only the metagenomes labelled as `sewage` (i.e., not the time series metagenomes).
2.  This will take at least 2 hours so be sure to get through this step even if you don't plan to immediately proceed further.
3.  When finished, you should expect to find a FASTA file of the assembly in each of the output directory called `contigs.fasta`. Ensure you have this file before proceeding -- if it's missing, `SPAdes` may have failed so check your log files and react accordingly.

## **Step 3: Binning for metagenome assembled genomes (MAGs)**
1. Use `step03.sbatch` to bin the contigs generated by `SPAdes` in step 2.
2. Copy and rename all of the bins produced by `maxbin` and `metabat`.
```
for genome in 03_bins/sewage[ABC]/*.fa*; do sample=$(echo ${genome} | cut -f 2 -d "/"); str=$(basename ${genome} | cut -f 1-2 -d "."); mv ${genome} 03_bins/${sample}_${str}.fna; done
```
## **Step 4: MAG quality control and annotation**
1. Use `step04_checkm2.sbatch` to prepare a report on the quality of MAGs you retrieved (output as: `04_annotations/checkm2/quality_report.tsv`)
2. Calculate the quality score (Q) of each MAG as Q = Completeness - 5 * Contamination
3. Remove any MAGs from further analysis with Q < 50
4. Compare the remaining MAGs all v. all using `fastANI` (recall the tool's matrix option)
5. For MAGs that match above 95% ANI, select as a representative for the pair whichever MAG has better quality while removing the other
6. Proceed with the remaining MAGs to annotation with `step04_bakta.sbatch`

## **Step 5: Taxonomic classification** 
1. Using only the MAGs passing step 4, prepare a batch file like so:
```
path/to/genome1.fna   genome1
path/to/genome2.fna   genome2
...
path/to/genomeN.fna   genomeN
```
2. Use `step05.sbatch` to taxonomically classify the MAGs in your batch file.

## **Step 6: Calculate relative abundances**
1. Use `step06.sbatch` to calculate the abundance of your MAGs across the dataset.
   
# Results

1.  List the names of the tools we used in this exercise and a link to their documentation.
2.  What was the total size of your assembly as reported by `SPAdes`? How many contigs did it contain?
3.  How many MAGs were reported by `maxbin` and `metabat`? What was the total count of MAGs with Q >= 50? Were there any redundant MAGs?
4.  What fraction of the assembly was binned into a good quality representative genome (i.e., those passing step 4)?
5.  Prepare a table summarizing the taxonomic classification of each MAG in your final collection (i.e., those MAGs passing step 4 and input to step 5).
6.  Prepare a table showing which MAGs appear to contain genes encoding oxidase and catalase. 
7.  Prepare a visual showing the abundances of each of these MAGs in the `sewage` sample and across the time series samples.

# Discussion

1.  Why did we use two different binning software in this exercise? Explain what is meant by MAG dereplication and how we accomplished this in our exercise.
2.  Define "single copy marker genes" in the context of measuring genome completeness and contamination. Provide examples of these types of genes.
3.  What taxonomic database did we classify our MAGs with? What gene database(s) did `bakta` use for annotating our genomes?
4.  Approximately how long does it take for the total sewage-associated signal to fade in the time series?
5.  Did the populations represented by your MAGs share similar decay patterns? If they were different, hypothesize why that might be.
6.  Do any of the MAGs carry antimicrobial resistance genes? If so, which types of AMR genes do they carry and what is the persistence of those AMR-carrying populations in the environment?
